src/libsyntax/parse/token.rs

```diff
@@ -203,6 +233,11 @@ impl Token {
         Token::Interpolated(Lrc::new((nt, LazyTokenStream::new())))
     }
 
+    /// Recovers a `Token` from an `ast::Ident`. This creates a raw identifier if necessary.
+    pub fn from_ast_ident(ident: ast::Ident) -> Token {
+        Ident(ident, is_reserved_ident(ident))
```

Comment:
This function and all its uses are very suspicious.
It's used when we previously lost the "rawness" property previously and then we are trying to guess it - if it's a keyword, then it was a raw identifier otherwise not. We shouldn't do this guessing, we need to keep the `bool` instead or use the `"r#{}"` string encoding like in proc macros if keeping `bool` is not possible for some reason.

Reply:
This problem exists when we go from tokens to AST (including `MetaItem`) and then have to back to tokens.

Reply:
Where in the AST should this information be encoded? `MetaItem` can be handled with a new struct, but in the other use of this method was `impl ToTokens for ast::Ident` (how is `ToToken` used anyway?). In the issue, you mentioned that ast::Ident would be a bad place for an `is_raw` parameter, because of how size could be an issue for it.

Reply:
I tend to leave this as is for now, it's possible that we won't have this roundtrip "Tokens -> AST -> Tokens" in the future and will just keep the original tokens.
I need to think about this a bit more.